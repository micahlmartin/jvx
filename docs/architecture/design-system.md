# Design System Documentation

## Overview
This document outlines the core design principles and specifications for our application's components, with a particular focus on node visualization and interaction elements.

## Handle Components

### Visual Properties
- **Size**: 
  - Width: ${designSystem.handles.size.width}
  - Height: ${designSystem.handles.size.height}
- **Shape**: Circular (border-radius: 50%)
- **Border**: None
- **Colors**:
  - Default: ${designSystem.handles.colors.default}
  - Hover: ${designSystem.handles.colors.hover}

### Positioning
- **Left Handle**:
  - Position: Absolute
  - Left: 0
  - Top: 50%
  - Transform: translate(-50%, -50%)
  
- **Right Handle**:
  - Position: Absolute
  - Right: 0
  - Top: 50%
  - Transform: translate(50%, -50%)

### Visibility Rules
1. **Target Handles**:
   - Default state: Invisible (opacity: 0)
   - When connecting: Visible (opacity: 1)
   - Pointer events: None when invisible, All when visible

2. **Source Handles**:
   - Default state: Visible (opacity: 1)
   - When not visible: Hidden (opacity: 0)
   - Pointer events: All when visible, None when hidden

### Transitions
- Opacity transition: ${designSystem.handles.transitions.opacity}
- Z-index: ${designSystem.zIndex.handles}

## Property Row Component

### Layout
- Display: Grid
- Grid Template: auto 1fr auto
- Alignment: Center
- Gap: 8px
- Minimum Height: 22px

### Separator
- Position: Absolute
- Color: ${designSystem.borders.separator}
- Location:
  - Left: -${designSystem.spacing.nodePadding}
  - Right: -${designSystem.spacing.nodePadding}
  - Bottom: -${parseInt(designSystem.spacing.propertyGap) / 2}px
- Height: 1px
- Visibility: Present on all rows except the last child

## Design Principles

### 1. Visual Hierarchy
- Handles should be visually distinct but not dominate the interface
- Property rows should maintain clear separation while preserving visual flow

### 2. Interaction Design
- Handle visibility should provide clear affordances for connection possibilities
- Property row separators should create clear visual boundaries without creating visual noise

### 3. Consistency
- All measurements should use the design system's spacing units
- Colors should be consistently applied using design system tokens
- Transitions should be smooth and consistent across similar interactions

## Usage Guidelines

### 1. Handle Implementation
- Always use the NodeHandle component for connection points
- Never override the default visibility behavior without explicit justification
- Maintain the positioning system to ensure consistent alignment

### 2. Property Row Implementation
- Always use the PropertyRow component for node properties
- Maintain the separator system for visual clarity
- Respect the grid layout system for consistent alignment

## Accessibility Considerations
- Ensure sufficient color contrast for all interactive elements
- Maintain appropriate hit areas for handle interactions
- Provide clear visual feedback for interactive states

## Node Structure

### Property Rows
- Must maintain consistent separation with borders
- Must use pseudo-element implementation for borders:
```css
&:not(:last-child)::after {
  content: '';
  position: absolute;
  left: -${designSystem.spacing.nodePadding};
  right: -${designSystem.spacing.nodePadding};
  bottom: -${parseInt(designSystem.spacing.propertyGap) / 2}px;
  height: 1px;
  background: ${designSystem.borders.separator};
}
```

### Spacing
- Node padding: ${designSystem.spacing.nodePadding}
- Property gap: ${designSystem.spacing.propertyGap}
- Header gap: ${designSystem.spacing.headerGap}

### Visual Hierarchy
- Node background: var(--node-bg)
- Border: ${designSystem.borders.node}
- Border radius: 8px
- Backdrop filter: blur(12px)

## Typography

### Font Sizes
- Type: ${designSystem.typography.sizes.type}
- Label: ${designSystem.typography.sizes.label}
- Property: ${designSystem.typography.sizes.property}

### Font Families
- Monospace: ${designSystem.typography.fontMono}
- Default system font for general text

## Colors

### Node Headers
- Level 0-4 backgrounds and text colors defined in designSystem.colors.nodeHeaders
- Colors cycle every 5 levels
- Special handling for array items and root nodes

### Property Values
- String: ${designSystem.colors.values.string}
- Number: ${designSystem.colors.values.number}
- Boolean: ${designSystem.colors.values.boolean}
- Default: ${designSystem.colors.values.default}

## Z-Index Hierarchy
- Base Node: ${designSystem.zIndex.nodeBase}
- Handles: ${designSystem.zIndex.handles}

## Transitions
- Handle opacity: ${designSystem.handles.transitions.opacity}
- Node hover effects: 0.2s ease

## Implementation Requirements

### Handle Components
1. Must use design system constants for positioning
2. Must maintain proper opacity settings
3. Must use absolute positioning
4. Must implement proper z-indexing

### Property Row Components
1. Must implement separator borders using pseudo-elements
2. Must maintain consistent spacing
3. Must follow typography guidelines
4. Must implement proper value type colors

### Node Components
1. Must follow visual hierarchy guidelines
2. Must implement proper padding and spacing
3. Must maintain proper z-index stacking
4. Must implement proper backdrop filters

## Testing Requirements
1. Visual regression tests must verify handle positioning
2. Tests must verify proper opacity settings
3. Tests must verify border implementations
4. Tests must verify color implementations

## Accessibility
1. Proper color contrast for all text elements
2. Sufficient touch targets for mobile interaction
3. Proper focus indicators for interactive elements
4. Proper ARIA attributes for interactive elements 