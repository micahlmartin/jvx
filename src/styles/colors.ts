export const colors = {
  editor: {
    bg: {
      DEFAULT: '#FFFFFF',
      dark: '#1E1E2E',
      highlight: {
        DEFAULT: 'rgba(0, 0, 0, 0.05)',
        dark: 'rgba(255, 255, 255, 0.1)'
      },
      selection: {
        DEFAULT: 'rgba(0, 0, 0, 0.1)',
        dark: 'rgba(255, 255, 255, 0.2)',
        inactive: {
          DEFAULT: 'rgba(0, 0, 0, 0.05)',
          dark: 'rgba(255, 255, 255, 0.1)'
        }
      }
    },
    text: {
      DEFAULT: '#1A1A1A',
      dark: '#E4E4E7',
      muted: {
        DEFAULT: '#475569',
        dark: '#94A3B8'
      },
      placeholder: {
        DEFAULT: 'rgba(0, 0, 0, 0.3)',
        dark: 'rgba(255, 255, 255, 0.3)'
      }
    },
    gutter: {
      bg: {
        DEFAULT: '#F8F9FA',
        dark: '#1E1E2E'
      },
      modified: '#3B82F6',
      added: '#10B981',
      deleted: '#EF4444'
    },
    error: {
      text: '#EF4444',
      bg: 'rgba(239, 68, 68, 0.1)'
    },
    bracket: {
      match: {
        bg: {
          DEFAULT: 'rgba(0, 0, 0, 0.1)',
          dark: 'rgba(255, 255, 255, 0.2)'
        },
        border: {
          DEFAULT: 'rgba(0, 0, 0, 0.2)',
          dark: 'rgba(255, 255, 255, 0.4)'
        }
      }
    }
  },
  node: {
    types: {
      root: {
        bg: 'rgba(82, 82, 91, 0.3)',
        text: 'rgba(161, 161, 170, 0.9)'
      },
      customer: {
        bg: 'rgba(59, 130, 246, 0.15)',
        text: 'rgba(96, 165, 250, 0.9)'
      },
      items: {
        bg: 'rgba(147, 51, 234, 0.15)',
        text: 'rgba(167, 139, 250, 0.9)'
      },
      shipping: {
        bg: 'rgba(236, 72, 153, 0.15)',
        text: 'rgba(236, 72, 153, 0.9)'
      },
      payment: {
        bg: 'rgba(16, 185, 129, 0.15)',
        text: 'rgba(52, 211, 153, 0.9)'
      },
      address: {
        bg: 'rgba(245, 158, 11, 0.15)',
        text: 'rgba(251, 191, 36, 0.9)'
      }
    },
    values: {
      string: '#38BDF8',
      number: '#EC4899',
      boolean: {
        true: '#4ADE80',
        false: '#F87171'
      },
      property: '#94A3B8'
    },
    border: {
      DEFAULT: 'rgba(255, 255, 255, 0.1)'
    }
  }
} as const; 