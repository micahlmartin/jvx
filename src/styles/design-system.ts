export const designSystem = {
  spacing: {
    nodePadding: '12px',
    propertyGap: '6px',
    headerGap: '8px',
  },

  typography: {
    fontMono: "'JetBrains Mono', monospace",
    sizes: {
      type: '12px',
      label: '14px',
      property: '13px',
    },
  },

  borders: {
    node: '1px solid var(--node-border)',
    separator: '1px solid var(--node-border)',
    radius: {
      node: '8px',
      badge: '4px',
    },
  },

  colors: {
    nodeHeaders: {
      level0: { background: '#2D3748', text: '#FFFFFF' },
      level1: { background: '#553C9A', text: '#FFFFFF' },
      level2: { background: '#2A4365', text: '#FFFFFF' },
      level3: { background: '#285E61', text: '#FFFFFF' },
      level4: { background: '#702459', text: '#FFFFFF' },
    },
    values: {
      string: 'var(--text-string)',
      number: 'var(--text-number)',
      boolean: 'var(--text-boolean)',
      default: 'var(--text-value)',
    },
  },

  handles: {
    size: {
      width: '8px',
      height: '8px',
    },
    colors: {
      default: 'var(--handle-color)',
      hover: 'var(--handle-hover-color)',
    },
    positioning: {
      source: {
        right: '0',
        top: '50%',
        transform: 'translate(50%, -50%)',
      },
      target: {
        left: '0',
        top: '50%',
        transform: 'translate(-50%, -50%)',
      },
    },
    opacity: {
      source: {
        visible: 1,
        hidden: 0,
      },
      target: {
        default: 0,
        connecting: 1,
      },
    },
    transitions: {
      opacity: 'opacity 0.2s ease',
    },
  },

  zIndex: {
    nodeBase: 1,
    handles: 10,
    tooltip: 100,
  },

  effects: {
    nodeBackdrop: 'blur(12px)',
    nodeHoverTransition: 'all 0.2s ease',
  },
} as const;

// Type exports for type safety
export type DesignSystem = typeof designSystem;
export type NodeHeaderLevel = keyof typeof designSystem.colors.nodeHeaders;
export type ValueType = keyof typeof designSystem.colors.values;
