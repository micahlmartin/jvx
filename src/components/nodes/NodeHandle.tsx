import React from 'react';
import { Handle, Position, useStore } from 'reactflow';
import styled from 'styled-components';
import { designSystem } from '../../styles/design-system';

/**
 * Props for the NodeHandle component.
 * @interface NodeHandleProps
 */
interface NodeHandleProps {
  /** The type of handle - either a source for outgoing connections or a target for incoming connections */
  type: 'source' | 'target';
  
  /** The position of the handle on the node - typically Left for targets and Right for sources */
  position: Position;
  
  /** Whether the handle is visible - only affects source handles as target handles have their own visibility logic */
  isVisible?: boolean;
  
  /** Optional ID for the handle - useful when multiple handles are present on the same node */
  id?: string;
  
  /** Optional className for additional styling */
  className?: string;
}

/**
 * Styled handle component that implements the design system specifications.
 * Handles the visual appearance and positioning of connection points.
 */
const StyledHandle = styled(Handle)<{ $isVisible: boolean; $isConnecting: boolean }>`
  position: absolute;
  width: ${designSystem.handles.size.width};
  height: ${designSystem.handles.size.height};
  background: ${designSystem.handles.colors.default};
  border: none;
  border-radius: 50%;
  cursor: pointer;
  opacity: ${({ $isVisible, $isConnecting, type }) => {
    if (type === 'target') return $isConnecting ? 1 : 0;
    return $isVisible ? 1 : 0;
  }};
  pointer-events: ${({ $isVisible, $isConnecting, type }) => {
    if (type === 'target') return $isConnecting ? 'all' : 'none';
    return $isVisible ? 'all' : 'none';
  }};
  transition: ${designSystem.handles.transitions.opacity};
  z-index: ${designSystem.zIndex.handles};

  &:hover {
    background: ${designSystem.handles.colors.hover};
  }

  ${({ position }) => {
    switch (position) {
      case Position.Left:
        return `
          left: 0;
          top: 50%;
          transform: translate(-50%, -50%);
        `;
      case Position.Right:
        return `
          right: 0;
          top: 50%;
          transform: translate(50%, -50%);
        `;
      default:
        return '';
    }
  }}
`;

/**
 * NodeHandle component that provides connection points for nodes in a flow diagram.
 * 
 * This component follows the design system specifications for handle appearance and behavior:
 * - Target handles are invisible by default and only show when a connection is being made
 * - Source handles are visible by default but can be hidden via the isVisible prop
 * - Handles are positioned exactly at the edge of nodes with a visual offset
 * - Transitions are smooth for all visibility changes
 * 
 * @component
 * @example
 * // Target handle on the left side of a node
 * <NodeHandle
 *   type="target"
 *   position={Position.Left}
 * />
 * 
 * @example
 * // Source handle on the right side of a node
 * <NodeHandle
 *   type="source"
 *   position={Position.Right}
 *   isVisible={true}
 * />
 */
export const NodeHandle: React.FC<NodeHandleProps> = ({ 
  type, 
  position, 
  isVisible = true, 
  id,
  className 
}) => {
  const isConnecting = useStore((state) => state.connectionNodeId != null);

  return (
    <StyledHandle
      type={type}
      position={position}
      $isVisible={isVisible}
      $isConnecting={isConnecting}
      id={id}
      className={className}
    />
  );
};

export default NodeHandle; 