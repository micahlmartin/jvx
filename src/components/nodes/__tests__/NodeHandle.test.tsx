import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';
import { Position } from 'reactflow';
import { NodeHandle } from '../NodeHandle';
import { designSystem } from '../../../styles/design-system';

describe('NodeHandle', () => {
  describe('Visibility Rules', () => {
    it('should render target handle with opacity 0 by default', () => {
      const { container } = render(
        <NodeHandle
          type="target"
          position={Position.Left}
        />
      );
      
      const handle = container.firstChild as HTMLElement;
      expect(handle).toHaveStyle({
        opacity: '0',
        pointerEvents: 'none'
      });
    });

    it('should render source handle with opacity 1 by default', () => {
      const { container } = render(
        <NodeHandle
          type="source"
          position={Position.Left}
        />
      );
      
      const handle = container.firstChild as HTMLElement;
      expect(handle).toHaveStyle({
        opacity: '1',
        pointerEvents: 'all'
      });
    });

    it('should render source handle with opacity 0 when not visible', () => {
      const { container } = render(
        <NodeHandle
          type="source"
          position={Position.Left}
          isVisible={false}
        />
      );
      
      const handle = container.firstChild as HTMLElement;
      expect(handle).toHaveStyle({
        opacity: '0',
        pointerEvents: 'none'
      });
    });
  });

  describe('Positioning', () => {
    it('should position handle on the left', () => {
      const { container } = render(
        <NodeHandle
          type="source"
          position={Position.Left}
        />
      );
      
      const handle = container.firstChild as HTMLElement;
      expect(handle).toHaveStyle({
        left: '0',
        top: '50%',
        transform: 'translate(-50%, -50%)'
      });
    });

    it('should position handle on the right', () => {
      const { container } = render(
        <NodeHandle
          type="source"
          position={Position.Right}
        />
      );
      
      const handle = container.firstChild as HTMLElement;
      expect(handle).toHaveStyle({
        right: '0',
        top: '50%',
        transform: 'translate(50%, -50%)'
      });
    });
  });

  describe('Styling', () => {
    it('should use correct size from design system', () => {
      const { container } = render(
        <NodeHandle
          type="source"
          position={Position.Right}
        />
      );
      
      const handle = container.firstChild as HTMLElement;
      expect(handle).toHaveStyle({
        width: designSystem.handles.size.width,
        height: designSystem.handles.size.height
      });
    });

    it('should use correct colors from design system', () => {
      const { container } = render(
        <NodeHandle
          type="source"
          position={Position.Right}
        />
      );
      
      const handle = container.firstChild as HTMLElement;
      expect(handle).toHaveStyle({
        background: designSystem.handles.colors.default
      });
    });
  });

  describe('Interaction Behavior', () => {
    it('should have pointer-events: none when target handle and not connecting', () => {
      const { container } = render(
        <NodeHandle
          type="target"
          position={Position.Left}
        />
      );
      
      const handle = container.firstChild as HTMLElement;
      expect(handle).toHaveStyle({
        'pointer-events': 'none'
      });
    });

    it('should have pointer-events: all when source handle and visible', () => {
      const { container } = render(
        <NodeHandle
          type="source"
          position={Position.Right}
          isVisible={true}
        />
      );
      
      const handle = container.firstChild as HTMLElement;
      expect(handle).toHaveStyle({
        'pointer-events': 'all'
      });
    });
  });

  describe('Visual Regression', () => {
    it('should match snapshot for target handle', () => {
      const { container } = render(
        <NodeHandle
          type="target"
          position={Position.Left}
        />
      );
      
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for source handle', () => {
      const { container } = render(
        <NodeHandle
          type="source"
          position={Position.Right}
          isVisible={true}
        />
      );
      
      expect(container.firstChild).toMatchSnapshot();
    });
  });
}); 