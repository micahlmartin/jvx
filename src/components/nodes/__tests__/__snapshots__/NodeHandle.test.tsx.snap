// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeHandle Visual Regression should match snapshot for source handle 1`] = `
.c0 {
  position: absolute;
  width: 8px;
  height: 8px;
  background: var(--handle-color);
  border: none;
  border-radius: 50%;
  cursor: pointer;
  opacity: 1;
  pointer-events: all;
  transition: opacity 0.2s ease;
  z-index: 10;
  left: 0;
  top: 50%;
  transform: translate(-50%, -50%);
}

.c0:hover {
  background: var(--handle-hover-color);
}

<div
  class="c0"
  data-testid="handle-source"
  style="opacity: 1; pointer-events: all;"
/>
`;

exports[`NodeHandle Visual Regression should match snapshot for target handle 1`] = `
.c0 {
  position: absolute;
  width: 8px;
  height: 8px;
  background: var(--handle-color);
  border: none;
  border-radius: 50%;
  cursor: pointer;
  opacity: 1;
  pointer-events: all;
  transition: opacity 0.2s ease;
  z-index: 10;
  left: 0;
  top: 50%;
  transform: translate(-50%, -50%);
}

.c0:hover {
  background: var(--handle-hover-color);
}

<div
  class="c0"
  data-testid="handle-target"
  style="opacity: 0; pointer-events: none;"
/>
`;
